File: js/lib/graph.js
/**
 * ESTRUTURA DE DADOS GRAFO
 * 
 * É uma estrutura de dados não-linear, formada por vértices (ou nodos) e arestas (ou arcos).
 * Sua principal finalidade é representar as relações entre diferentes objetos.
 * Tais relações podem ser bidirecionais, resultando em grafos não direcionados, ou unidirecionais, constituindo grafos direcionados (digrafos).
 * Tem várias aplicações, como a representação de redes de computadores, mapas e caminhos e redes sociais.
 */
class Graph {

    /** Método construtor */
    constructor(isDirected = false) {
        this.isDirected = isDirected;
        this.data = {}; // Objeto vazio para armazenar o grafo no formato de lista de adjacência
    }

    /** Método que adiciona um vértice ao grafo */
    addVertex(val) {
        if (!(val in this.data)) {
            this.data[val] = new Set(); // Conjunto vazio
        }
    }

    /** 
     * Método que adiciona uma aresta entre dois vértices
     * 
     * origin: vértice de origem
     * dest: vértice de destino
     * rel: nome do relacionamento (opcional)
     */
    addEdge(origin, dest, rel = null) {
        if (!(origin in this.data)) {
            this.addVertex(origin);
        }
        if (!(dest in this.data)) {
            this.addVertex(dest);
        }

        const edge1 = [dest, rel];

        this.data[origin].add(edge1);

        if (!this.isDirected) {
            const edge2 = [origin, rel];
            this.data[dest].add(edge2);
        }
    }

    /** Método que retorna o grau (número de vértices adjacentes) de um vértice */
    degree(vertex) {
        let result = 0;

        // Grau de saída (out-degree)
        if (vertex in this.data) {
            result = this.data[vertex].size;
        }

        // Grau de entrada (in-degree)
        for (const v in this.data) {
            for (const e of this.data[v]) {
                if (vertex === e[0]) {
                    result++;
                }
            }
        }

        return result;
    }

    /** 
     * Método que exclui um vértice do grafo
     * (Para que um vértice possa ser excluído, ele não pode ter arestas incidentes, ou seja, seu grau deve ser 0, nem ser referenciado por outros vértices)
     */
    removeVertex(vertex) {
        if (this.degree(vertex) !== 0) {
            throw new Error(
                "ERRO: impossível remover um vértice com arestas incidentes a ele."
            );
        } else if (vertex in this.data) {
            delete this.data[vertex];
        }
    }

    /** Método que remove uma aresta do grafo */
    removeEdge(origin, dest, rel = null) {
        const edge1 = [dest, rel];

        if (this.data[origin]) {
            this.data[origin].forEach((edge, index) => {
                if (edge[0] === dest && edge[1] === rel) {
                    this.data[origin].delete(edge);
                }
            });
        }

        if (!this.isDirected && this.data[dest]) {
            this.data[dest].forEach((edge, index) => {
                if (edge[0] === origin && edge[1] === rel) {
                    this.data[dest].delete(edge);
                }
            });
        }
    }

    /** Método que representa o grafo como uma string */
    toString() {
        let output = '';
        for (const vertex in this.data) {
            output += `${vertex}:\n`;
            this.data[vertex].forEach((edge, index) => {
                const [dest, rel] = edge;
                output += `  ${dest} (${rel})\n`;
            });
            output += '\n';
        }
        output += `isDirected: ${this.isDirected}\n`;
        return output;
    }
}

module.exports = Graph;

File: prova_2/turma_b/lib/graph.py
class Graph:
    """ ESTRUTURA DE DADOS GRAFO

    É uma estrutura de dados não-linear, formada por vértices (ou nodos) e arestas (ou arcos).
    Sua principal finalidade é representar as relações entre diferentes objetos.
    Tais relações podem ser bidirecionais, resultando em grafos não direcionados, ou unidirecionais, constituindo grafos direcionados (digrafos).
    Tem várias aplicações, como a representação de redes de computadores, mapas e caminhos e redes sociais. """

    def __init__(self, is_directed = False):
        """ Método construtor """
        self.__is_directed = is_directed

        # __data armazenará o grafo no formato de LISTA DE ADJACÊNCIA
        self.__data = {}    # Dicionário vazio

    def add_vertex(self, val):
        """ Método que adiciona um vértice ao grafo """
        # Verifica se o vértice já existe no dicionário. Só cria se o vértice não existir ainda
        if not val in self.__data:
            self.__data[val] = set()    # Conjunto vazio

    def add_edge(self, origin, dest, rel = None):
        """ Método que adiciona uma aresta entre dois vértices

        origin: vértice de origem
        dest: vértice de destino
        rel: nome do relacionamento (opcional) """

        # Cria os vértices de origem e destino, caso não existam ainda
        if not origin in self.__data: self.add_vertex(origin)
        if not dest in self.__data: self.add_vertex(dest)

        # Monta a aresta
        edge1 = (dest, rel)     # Isto é uma tupla ("lista imutável")

        # Adiciona a relação origem->destino
        self.__data[origin].add(edge1)

        # Se o grafo não for direcionado, adicionamos também a relação no sentido oposto, isto é, destino->origem
        if not self.__is_directed:
            edge2 = (origin, rel)   # Tupla
            self.__data[dest].add(edge2)

    def degree(self, vertex):
        """ Método que retorna o grau (número de vértices adjacentes) de um vértice """

        result = 0
        # GRAU DE SAÍDA (OUT-DEGREE)
        # É determinado contando-se o número de arestas associadas à entrada de dicionário correspondente ao vértice
        if vertex in self.__data: result = len(self.__data[vertex])

        # GRAU DE ENTRADA (IN-DEGREE)
        # É determinado procurando-se referências ao vértice nas arestas associadas a todos os vértices do dicionário
        for v in self.__data.keys():    # Percorre o dicionário (vértices)
            for e in self.__data[v]:    # Percorre os conjuntos (arestas)
                if vertex == e[0]: result += 1

        # O grau final é a soma dos graus de entrada e de saída
        return result

    def remove_vertex(self, vertex):
        """ Método que exclui um vértice do grafo
        (Para que um vértice possa ser excluído, ele não pode ter arestas incidentes, ou seja, seu grau deve ser 0, nem ser referenciado por outros vértices) """

        if self.degree(vertex) != 0:
            raise Exception("ERRO: impossível remover um vértice com arestas incidentes a ele.")
        # Vértice de grau 0, podemos excluir, caso exista no dicionário
        elif vertex in self.__data: del self.__data[vertex]

    def remove_edge(self, origin, dest, rel = None):
        """ Método que remove uma aresta do grafo """
        
        edge1 = (dest, rel)
        # Retira a tupla edge1 do conjunto de arestas do vértice de origem
        self.__data[origin].discard(edge1)

        # Se o grafo não for direcionado, precisamos remover também a aresta em sentido contrário
        if not self.__is_directed:
            edge2 = (origin, rel)
            self.__data[dest].discard(edge2)

    def __str__(self):
        """ Método que representa o grafo como uma string """
        output = str(self.__data)
        output += f"\nis_directed: {self.__is_directed}\n\n"
        return output
